atlite/data.py

```diff
@@ -0,0 +1,215 @@
+        elif isinstance(node, ast.Tuple):
+            return tuple(map(_convert, node.elts))
+        elif isinstance(node, ast.List):
+            return list(map(_convert, node.elts))
+        elif isinstance(node, ast.Set):
+            return set(map(_convert, node.elts))
+        elif isinstance(node, ast.Dict):
+            return dict((_convert(k), _convert(v))
+                        for k, v in zip(node.keys, node.values))
+        elif isinstance(node, ast.NameConstant):
+            return node.value
+        elif isinstance(node, ast.UnaryOp) and isinstance(node.op, (ast.UAdd, ast.USub)):
+            operand = _convert(node.operand)
+            if isinstance(operand, (int, float, complex)):
+                if isinstance(node.op, ast.UAdd):
+                    return + operand
+                else:
+                    return - operand
+        elif isinstance(node, ast.Call) and node.func.id == 'slice':
+            return slice(*map(_convert, node.args))
+
+        raise ValueError('malformed creation parameters: ' + repr(node))
+
+    return _convert(node_or_string)
+
+def _get_creation_parameters(data):
+    return literal_eval_creation_parameters(data.attrs['creation_parameters'])
+
+def requires_coords(f):
+    @wraps(f)
+    def wrapper(cutout, *args, **kwargs):
+        if not cutout.data.coords:
+            creation_parameters = _get_creation_parameters(cutout.data)
+            cutout.data = cutout.data.merge(cutout.dataset_module.get_coords(**creation_parameters))
+        return f(cutout, *args, **kwargs)
+    return wrapper
+
+
+class requires_windowed(object):
+    def __init__(self, features, windows=None, allow_dask=False):
+        self.features = features
+        self.windows = windows
+        self.allow_dask = allow_dask
+
+    def __call__(self, f):
+        @wraps(f)
+        def wrapper(cutout, *args, **kwargs):
+            features = kwargs.pop('features', self.features)
+            window_type = kwargs.pop('windows', self.windows)
+            windows = create_windows(cutout, features, window_type, self.allow_dask)
+
+            return f(cutout, *args, windows=windows, **kwargs)
+
+        return wrapper
+
+def create_windows(cutout, features, window_type, allow_dask):
+    features = set(features if features is not None else cutout.available_features)
+    missing_features = features - set(cutout.data.attrs.get('prepared_features', []))
+
+    if missing_features:
+        logger.error(f"The following features need to be prepared first: {', '.join(missing_features)}. Will try anyway!")
+
+    if window_type is False:
+        return [cutout.data]
+    else:
+        return Windows(cutout, features, window_type, allow_dask)
+
+class Windows(object):
+    def __init__(self, cutout, features, window_type=None, allow_dask=False):
+        group_kws = {}
+        if window_type is None:
+            group_kws['grouper'] = pd.Grouper(freq="M")
+        elif isinstance(window_type, base_string):
```

Comment:
```suggestion
        elif isinstance(window_type, str):
```

as `base_string` is no longer available in Python 3.X.